### PR TITLE
QUIC wire format support

### PR DESCRIPTION
**Pull request by hlandau**

### CODE REVIEW

File: include/internal/quic_vlint.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * integer encoding.
+ */
+size_t ossl_quic_vlint_encode_len(uint64_t v);
+
+/*
+ * This function writes a QUIC varable-length encoded integer to buf.
+ * The smallest usable representation is used.
+ *
+ * It is the caller's responsibility to ensure that the buffer is big enough by
+ * calling ossl_quic_vlint_enc_len(v) before calling this function.
```
Comment (mattcaswell) [inline include/internal/quic_vlint.h:33]: ossl_quic_vlint_encode_len

File: include/internal/quic_vlint.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * Precondition: buf is at least ossl_quic_vlint_dec_len(buf[0]) bytes in size
+ *   (unchecked)
+ */
+uint64_t ossl_quic_vlint_decode_unchecked(const unsigned char *buf);
+
+/*
+ * Given a buffer buf of buf_len bytes in length, attempts to decode an encoded
+ * QUIC variable-length integer at the start of the buffer and writes the result
+ * to *v. If buf_len is inadequate, suggesting a truncated encoded integer, the
```
Comment (mattcaswell) [inline include/internal/quic_vlint.h:81]: What happens if `buf_len` is too much, i.e. there are trailing bytes in the buffer. Does this fail? Should it?

File: include/internal/quic_param.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# define QUIC_TRANSPORT_PARAM_INITIAL_MAX_STREAM_DATA_UNI                0x07
+# define QUIC_TRANSPORT_PARAM_INITIAL_MAX_STREAMS_BIDI                   0x08
+# define QUIC_TRANSPORT_PARAM_INITIAL_MAX_STREAMS_UNI                    0x09
+# define QUIC_TRANSPORT_PARAM_ACK_DELAY_EXPONENT                         0x0A
+# define QUIC_TRANSPORT_PARAM_MAX_ACK_DELAY                              0x0B
+# define QUIC_TRANSPORT_PARAM_DISABLE_ACTIVE_MIGRATION                   0x0C
+# define QUIC_TRANSPORT_PARAM_PREFERRED_ADDRESS                          0x0D
+# define QUIC_TRANSPORT_PARAM_ACTIVE_CONNECTION_ID_LIMIT                 0x0E
+# define QUIC_TRANSPORT_PARAM_INITIAL_SOURCE_CONNECTION_ID               0x0F
+# define QUIC_TRANSPORT_PARAM_RETRY_SOURCE_CONNECTION_ID                 0x10
```
Comment (mattcaswell) [inline include/internal/quic_param.h:30]: I am left wondering why this exists in this PR. These values don't seem to be used anywhere.

File: include/internal/packet.h [hunk 4]
```diff
@@ -594,6 +617,32 @@ __owur static ossl_inline int PACKET_get_length_prefixed_3(PACKET *pkt,
...
+ * buffer with the original |pkt|, so data wrapped by |pkt| must outlive the
+ * |subpkt|. Upon failure, the original |pkt| and |subpkt| are not modified.
+ */
+__owur static ossl_inline int PACKET_get_quic_length_prefixed(PACKET *pkt,
+                                                              PACKET *subpkt)
+{
+
+    uint64_t length;
+    const unsigned char *data;
+    PACKET tmp = *pkt;
```
Comment (mattcaswell) [inline include/internal/packet.h:633]: Add a blank line after declarations

File: crypto/packet.c [hunk 5]
```diff
@@ -510,3 +527,81 @@ void WPACKET_cleanup(WPACKET *pkt)
...
+    size_t enclen = ossl_quic_vlint_encode_len(v);
+
+    if (WPACKET_allocate_bytes(pkt, enclen, &b) == 0)
+        return 0;
+
+    ossl_quic_vlint_encode(b, v);
+    return 1;
+}
+
+int WPACKET_quic_allocate_transport_param_bytes(WPACKET *pkt, uint64_t id,
```
Comment (mattcaswell) [inline crypto/packet.c:575]: These transport_param functions seem oddly specific to a particular part of QUIC. Do we expect the writing of transport params to be sufficiently widespread to justify having specific support for them? If so, why do we not see equivalent functions in the PACKET API?

File: crypto/quic_vlint.c [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#include "internal/quic_vlint.h"
+#include "internal/e_os.h"
+
+size_t ossl_quic_vlint_encode_len(uint64_t v)
+{
+    if (v < 64)
```
Comment (mattcaswell) [inline crypto/quic_vlint.c:6]: Isn't this just `OSSL_QUIC_VLINT_2B_MIN`?

File: crypto/quic_vlint.c [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        return 1;
+
+    if (v < 16384)
+        return 2;
+
+    if (v < 1073741824)
```
Comment (mattcaswell) [inline crypto/quic_vlint.c:12]: `OSSL_QUIC_VLINT_8B_MIN`?

File: crypto/quic_vlint.c [hunk 8]
```diff
@@ -0,0 +1,91 @@
+#include "internal/quic_vlint.h"
+#include "internal/e_os.h"
+
+size_t ossl_quic_vlint_encode_len(uint64_t v)
+{
+    if (v < 64)
+        return 1;
+
+    if (v < 16384)
```
Comment (mattcaswell) [inline crypto/quic_vlint.c:9]: `OSSL_QUIC_VLINT_4B_MIN`

File: crypto/quic_vlint.c (hunk 6 above)
Comment (t-j-h) [inline crypto/quic_vlint.c:6]: OSSL_QUIC_VLINT_2B_MIN

File: crypto/quic_vlint.c (hunk 8 above)
Comment (t-j-h) [inline crypto/quic_vlint.c:9]: OSSL_QUIC_VLINT_4B_MIN

File: crypto/quic_vlint.c (hunk 7 above)
Comment (t-j-h) [inline crypto/quic_vlint.c:12]: OSSL_QUIC_VLINT_8B_MIN

File: test/packettest.c [hunk 9]
```diff
@@ -465,6 +465,110 @@ static int test_PACKET_as_length_prefixed_2(void)
...
+        { {0xc0,0x00,0x00,0x00,0x00,0x00,0x00,0x00}, 8, 0 },
+        { {0xc0,0x00,0x00,0x00,0x00,0x00,0x01,0x02}, 8, 258 },
+        { {0xfd,0x1f,0x59,0x8d,0xc9,0xf8,0x71,0x8a}, 8, 4404337426105397642 },
+    };
+
+    PACKET pkt;
+    size_t i;
+    uint64_t v;
+
+    for (i=0; i<OSSL_NELEM(cases); ++i) {
```
Comment (mattcaswell) [inline test/packettest.c:503]: Spaces around `=` and `<`

File: test/packettest.c [hunk 10]
```diff
@@ -465,6 +465,110 @@ static int test_PACKET_as_length_prefixed_2(void)
...
+
+        /* failure cases */
+        { {0x10}, 1, 0, 1 },
+        { {0x3f}, 1, 0, 1 },
+    };
+
+    size_t i;
+    PACKET pkt, subpkt = {0};
+
+    for (i=0; i<OSSL_NELEM(cases); ++i) {
```
Comment (mattcaswell) [inline test/packettest.c:544]: Spaces around `=` and `<`

File: include/internal/quic_param.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
+/*
```
Comment (t8m) [inline include/internal/quic_param.h:1]: This file does not seem to belong to this PR as these constants do not look relevant for the wire format support.

File: test/wpackettest.c [hunk 12]
```diff
@@ -424,6 +442,135 @@ static int test_WPACKET_init_der(void)
...
+            || !TEST_false(WPACKET_close(&pkt))
+                /* Now a top level so finish should succeed */
+            || !TEST_true(WPACKET_finish(&pkt))
+            || !TEST_true(WPACKET_get_total_written(&pkt, &written))
+            || !TEST_mem_eq(buf->data, written, quic1, sizeof(quic1)))
+        return cleanup(&pkt);
+
+    /* QUIC sub-packet with 1-byte length prefix, containing a 1-byte vlint */
+    if (!TEST_true(WPACKET_init(&pkt, buf))
+            || !TEST_true(WPACKET_start_quic_sub_packet_bound(&pkt, OSSL_QUIC_VLINT_2B_MIN-1))
```
Comment (mattcaswell) [inline test/wpackettest.c:468]: Spaces around `-`

File: test/wpackettest.c [hunk 13]
```diff
@@ -26,6 +26,24 @@ static const unsigned char simpleder[] = {
     0xfc, 0x04, 0x00, 0x01, 0x02, 0x03, 0xff, 0xfe, 0xfd
 };
 
+static const unsigned char quic1[] = { 0x80, 0x00, 0x00, 0x01, 0x09 };
```
Comment (t-j-h) [inline test/wpackettest.c:29]: It would be useful to have the comment about what each of these represent included here rather than just in the test_WPACKET_quic area - i.e. both places helps keep the context clear IMHO

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: include/internal/quic_param.h (hunk 11 above)
Comment (hlandau) [reply include/internal/quic_param.h:1]: It's intended for use with the transport parameter encoding functions, but you're correct that we could do without this for now.

File: include/internal/quic_vlint.h (hunk 2 above)
Comment (hlandau) [reply include/internal/quic_vlint.h:81]: I don't see why it should fail. I would expect the function to ordinarily be applied where there is a current position pointer and remaining length variable, which are progressively advanced and decreased respectively.

On second thoughts, I've changed `ossl_quic_vlint_decode` to return the number of bytes consumed in the success case, to make this use case easier.

File: include/internal/quic_param.h (hunk 3 above)
Comment (hlandau) [reply include/internal/quic_param.h:30]: These are used as the `id` param to `WPACKET_quic_write_transport_param_bytes/int`.

File: crypto/packet.c (hunk 5 above)
Comment (hlandau) [reply crypto/packet.c:575]: Makes sense. I'll remove these.

**Comment by hlandau** [comment]
Updated.

File: crypto/packet.c [hunk 14]
```diff
@@ -510,3 +527,54 @@ void WPACKET_cleanup(WPACKET *pkt)
...
+    pkt->subs->flags |= WPACKET_FLAGS_QUIC_VLINT;
+    return 1;
+}
+
+int WPACKET_start_quic_sub_packet(WPACKET *pkt)
+{
+    /*
+     * Assume no (sub)packet will exceed 4GiB, thus the 8-byte encoding need not
+     * be used.
+     * */
```
Comment (t8m) [inline crypto/packet.c:550]: No double `* *`

File: crypto/packet.c [hunk 15]
```diff
@@ -510,3 +527,54 @@ void WPACKET_cleanup(WPACKET *pkt)
...
+    return WPACKET_start_quic_sub_packet_bound(pkt, OSSL_QUIC_VLINT_4B_MIN);
+}
+
+int WPACKET_quic_sub_allocate_bytes(WPACKET *pkt, size_t len, unsigned char **allocbytes)
+{
+    if (   !WPACKET_start_quic_sub_packet_bound(pkt, len)
```
Comment (t8m) [inline crypto/packet.c:556]: Why these extra spaces after `(` ?

File: crypto/quic_vlint.c [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    if (v <= OSSL_QUIC_VLINT_MAX)
+        return 8;
+
+    return 0;
+}
+
+size_t ossl_quic_vlint_decode_len(uint8_t first_byte)
+{
+    return 1U << ((first_byte & 0xC0) >> 6);
+}
```
Comment (t8m) [inline crypto/quic_vlint.c:24]: Should these be static inline functions?

**Comment by mattcaswell** [comment]
Looks good - subject to @t8m's comments being resolved.

**Comment by hlandau** [comment]
Updated.

File: include/internal/packet.h [hunk 17]
```diff
@@ -228,6 +229,28 @@ __owur static ossl_inline int PACKET_peek_net_4(const PACKET *pkt,
...
+ */
+__owur static ossl_inline int PACKET_get_quic_vlint(PACKET *pkt,
+                                                    uint64_t *data)
+{
+    size_t enclen;
+
+    if (PACKET_remaining(pkt) < 1)
+        return 0;
+
+    enclen = ossl_quic_vlint_decode_len(*pkt->curr);
```
Comment (t8m) [inline include/internal/packet.h:244]: You'll need to add cast here to fix the windows build failure. Maybe with some assert or check that the decoded len < SIZE_T_MAX.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you.
